### PR TITLE
 hubble/metrics: Add support for fallback labels, ip addresses and dns names

### PR DESCRIPTION
**Pull request by gandro**
This adds a two context option to the Hubble metrics, `dns` and `ip`.

`dns` adds the DNS name as the source or destination label. If multiple DNS names are associated with an IP address, all of them are added.
`ip` adds the IPv4 or IPv6 address as the source or destination label.

In addition, this PR also fixes some missing cases around policy verdict and capture events. It also add support for specifying multiple contexts to support fall-back. For example, a metric configuration of `flow:destinationContext=dns|ip` will first try to use the DNS name of the target for the label, but if no DNS name is known, it will fall back and use the IP address of the target instead.

Except from `--hubble-metrics=flow:destinationContext=dns|ip` against the `connectivity-check.yaml`:

```
hubble_flows_processed_total{destination="1.1.1.1",protocol="TCP",subtype="to-stack",type="Trace",verdict="FORWARDED"} 42
hubble_flows_processed_total{destination="www.google.com",protocol="TCP",subtype="to-stack",type="Trace",verdict="FORWARDED"} 63

```

### CODE REVIEW

**Review by rolinh — APPROVED** [review]

**Review by qmonnet — APPROVED** [review]

**Comment by gandro** [comment]
test-me-please

**Comment by michi-covalent** [comment]
@gandro it might be useful to fall back to other context if dns name is not set? for example, if the destination is inside the cluster, i'd still like to see namespace / pod name, etc 🤔 not sure if it makes sense, but i was thinking something like:
```
--hubble-metrics=flow:destinationContext=namespace;prefer-dns=true
```
or something

**Comment by gandro** [comment]
> @gandro it might be useful to fall back to other context if dns name is not set? for example, if the destination is inside the cluster, i'd still like to see namespace / pod name, etc thinking not sure if it makes sense, but i was thinking something like:
> 
> ```
> --hubble-metrics=flow:destinationContext=namespace;prefer-dns=true
> ```
> 
> or something

I'd like to avoid making DNS a special case. All the other context options (except maybe identity) have the same problem in that the target might be empty. But maybe we can support a list of targets that are tried in sequence.

**Comment by gandro** [comment]
I have reworked the PR a bit, please see the PR description for details. @michi-covalent your use-case should now be supported by the following syntax:

```
--hubble-metrics=flow:destinationContext=dns|namespace
```

File: pkg/hubble/metrics/api/context.go [hunk 1]
```diff
@@ -151,6 +187,20 @@ func sourcePodShortContext(flow *pb.Flow) (context string) {
 	return
 }
 
+func sourceDNSContext(flow *pb.Flow) (context string) {
+	if flow.GetSourceNames() != nil {
+		context = strings.Join(flow.GetSourceNames(), ",")
```
Comment (michi-covalent) [inline pkg/hubble/metrics/api/context.go:192]: do you need to sort the dns name list, or is it already guaranteed to be sorted?

**Review by michi-covalent — APPROVED** [review]
🚀 lgtm a big 👍 just one question regarding the dns name field.

File: pkg/hubble/metrics/api/context.go (hunk 1 above)
Comment (gandro) [reply pkg/hubble/metrics/api/context.go:192]: Thanks for reminding me to check this, they are guaranteed to be sorted: https://github.com/cilium/cilium/blob/87b56a42a56633f1448c0c99a75dc26e5bb4c29f/pkg/fqdn/cache.go#L467

**Comment by gandro** [comment]
test-me-please

**Review by rolinh — APPROVED** [review]

**Pull request merged**